### PR TITLE
Use ConditionalPut when updating the RangeDescriptor.

### PR DESCRIPTION
**Pull request by bdarnell**
This primarily defends against the possibility noted in #768 in which
the former members of a range could revive and try to hijack it.

### CODE REVIEW

**Comment by tschottdorf** [comment]
LGTM. The issue previously was that a "slow" update could wreak havoc had a descriptor update happened in the meantime?

**Comment by bdarnell** [comment]
The issue is subtle:
* A range is initially present on nodes A, B, and C.
* Node C goes away for a while (partition, hardware failure, whatever).
* While it is gone, not only is node C replaced but the range is rebalanced away from all three nodes onto nodes D, E, and F.
* A and B clean up their local state from the range.
* When C comes back up, it is a follower in this range. It tries to talk to A and B, who have forgotten everything about this range so it is recreated on demand. The three nodes form a new quorum and believe themselves to be the valid members of the range!
* If the rogue quorum were allowed to perform a range descriptor update, they could redirect traffic from the legitimate copy of the range to themselves.

It's actually worse than this because range descriptors are cached and updated lazily. A client/gateway with a stale range descriptor might try to talk to the rogue nodes and believe everything is fine (these nodes could commit new commands among themselves)

**Comment by mrtracy** [comment]
LGTM

File: storage/range.go [hunk 1]
```diff
@@ -2219,6 +2221,37 @@ func (r *Range) ChangeReplicas(changeType proto.ReplicaChangeType, replica proto
 	return nil
 }
 
+// updateRangeDescriptor returns a client.Call to execute a ConditionalPut
+// on the range descriptor.
+func updateRangeDescriptor(descKey proto.Key, oldDesc, newDesc *proto.RangeDescriptor) client.Call {
```
Comment (petermattis) [inline storage/range.go]: Perhaps name this `updateRangeDescriptorCall` to follow the naming of the `client.*Call` functions.

File: storage/range.go [hunk 2]
```diff
@@ -2219,6 +2221,37 @@ func (r *Range) ChangeReplicas(changeType proto.ReplicaChangeType, replica proto
 	return nil
 }
 
+// updateRangeDescriptor returns a client.Call to execute a ConditionalPut
+// on the range descriptor.
```
Comment (spencerkimball) [inline storage/range.go]: This is really where some of the detail from the PR description would help people down the road quickly determine why the conditional put is necessary. How about:

// The conditional put verifies that changes to the range descriptor are made
// in a well-defined order, preventing a scenario where a wayward replica which
// is no longer part of the original Raft group comes back online to form a splinter
// group with a node which was also a former replica, and hijacks the range
// descriptor.

**Comment by spencerkimball** [comment]
LGTM

**Pull request merged**